sdc/datatypes/hpat_pandas_series_functions.py

```diff
@@ -445,66 +448,118 @@ def hpat_pandas_series_value_counts(self, normalize=False, sort=True, ascending=
         ty_checker.raise_exc(dropna, 'boolean', 'dropna')
 
     if isinstance(self.data, StringArrayType):
-        def hpat_pandas_series_value_counts_str_impl(self, normalize=False, sort=True, ascending=False, bins=None, dropna=True):
-            # TODO: if dropna add nan handling
+        def hpat_pandas_series_value_counts_str_impl(
+                self, normalize=False, sort=True, ascending=False, bins=None, dropna=True):
 
-            value_counts_dict = {}
+            value_counts_dict = Dict.empty(
+                key_type=types.unicode_type,
+                value_type=types.intp
+            )
 
-            for value in self._data:
-                if value in value_counts_dict:
-                    value_counts_dict[value] += 1
-                else:
-                    value_counts_dict[value] = 1
+            nan_counts = 0
+            for i, value in enumerate(self._data):
+                if str_arr_is_na(self._data, i):
+                    if not dropna:
+                        nan_counts += 1
+                    continue
+
+                value_counts_dict[value] = value_counts_dict.get(value, 0) + 1
+
+            need_add_nan_count = not dropna and nan_counts
 
-            # TODO: workaround, keys() result can not be casted to array type
-            # TODO: use list comprehension instead or self.unique()
-            unique_values = [key for key in value_counts_dict]
-            unique_values_len = len(unique_values)
+            values = [key for key in value_counts_dict]
+            counts_as_list = [value_counts_dict[key] for key in value_counts_dict.keys()]
```

Comment:
Are we able to use method `values()` directly?
`[val for val in value_counts_dict.values()]`

Reply:
Yes, but for some strange reason it produces different results. I'll investigate that separately.